OpenTracing Shim: Add Tracer.Close() operation. (#3048)
open-telemetry/opentelemetry-specification: specification/compatibility/opentracing.md

or no value could be extracted, depending on the specific OpenTracing Language API
(e.g. Go and Python do, but Java may not).

## Close

> Optional operation. If this operation is implemented for a specific OpenTracing language,

Reviewer: https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#notation-conventions-and-compliance

```suggestion
OPTIONAL operation. If this operation is implemented for a specific OpenTracing language,
```

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Reviewer 2: ```suggestion
it MUST close the underlying `Tracer` if it implements a "closeable" interface or method;
```

By my logic, it makes more sense to ask whether the OpenTracing `Tracer` implements a closeable interface, not the `TracerProvider`.

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Author: Oh, this is because OTel exposes this functionality on `TracerProvider`, not `Tracer` :(

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Author: Hey @jmacd - circling back, there was a [RFC](https://github.com/opentracing/specification/blob/master/rfc/closeable_tracer.md#specification-changes) in OpenTracing explaining this part:

> For Tracers that record Spans for persistance, calling Close should flush the current in-memory collection to the configured persistance store.

So not quite about reference counting, at least in the general purpose.

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Reviewer 2: Sounds good -- this leaves a question about the behavior of OpenTracing behavior. If I create two Tracers from the shim, then close them one at a time, does this mean the TracerProvider is shutdown twice?  If this sounds problematic (which it does to me), I think we should say that Close() calls `Flush()`, not `Shutdown()`, to protect against double close?

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Author: So the problem is that `Flush()` is not 'exposed' by any interface (such as `Closeable` in the Java world). So we'd need to either actually do a direct cast (e.g. to `TracerProviderSdk`) or do a full-duck type access (which in C++ and Java would require reflection instead)

I'd instead add a recommendation on NOT creating more than an OpenTracing `Tracer`  (which is exactly the typical use case in OT), as well a catching exceptions when calling `TracerProvider.Close()` from the Shim side, to at least provide a clean operation. What do you think?

> it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;

Reviewer 2: > I'd instead add a recommendation on NOT creating more than an OpenTracing Tracer (which is exactly the typical use case in OT)

Sounds good! Thanks @carlosalberto.

> OpenTelemetry API, and else it MUST be defined as a no-op operation.

Reviewer 3: ```suggestion
OpenTelemetry SDK, otherwise it MUST be defined as a no-op operation.
```

Author: The OpenTracing Shim is expected to NOT depend on the SDK, but on the API ;(

Reviewer 3: Then it seems like we're stuck, because OTEL API does not define Close/Flush equivalent.

Author: Right. So the options are a) We push for `TracerProvider.Close()` (or `Flush()`) or b) Have a no-op for now, waiting for this to be added in the future. Which one do you suggest we go after?

Reviewer 3: One other option is to allow lang-specific duck typing, eg if Java tracer provider implements a Closeable then cast and call it.

Author: @yurishkuro I prefer this last approach. Update the PR, let me know whether the current wording is clear enough.

or no value could be extracted, depending on the specific OpenTracing Language API
(e.g. Go and Python do, but Java may not).

## Close

> Optional operation. If this operation is implemented for a specific OpenTracing language,
> it MUST flush the underlying `TracerProvider` if such functionality is defined in the
> OpenTelemetry API, and else it MUST be defined as a no-op operation.

Reviewer 3: ```suggestion
```

Optional operation. If this operation is implemented for a specific OpenTracing language,

> it MUST close the underlying `TracerProvider` if it implements a closeable interface or method,
> and else it MUST be defined as a no-op operation.

Reviewer 3: ```suggestion
it MUST close the underlying `TracerProvider` if it implements a "closeable" interface or method;
otherwise it MUST be defined as a no-op operation.
```

or no value could be extracted, depending on the specific OpenTracing Language API
(e.g. Go and Python do, but Java may not).

## Close

Mark the `Tracer` Shim as closed, effectively preventing creation of new `Span` instances.

> Subsequent calls to [Start a new Span](#start-a-new-span) MUST return a no-op instance.

Reviewer 3: If I recall it was primarily used to force-flush any spans that were already finished but not exported (i.e. sitting in a memory buffer). Not starting new live spans was less of a goal (I don't even recall that one, and not sure why it would be important, because if those spans are not finished than they are not that different from no-op).

Author: That's correct, that operation was added to force-flush (before closing the app). Since we don't the equivalent in the OTel API, the alternative is that this call becomes a no-op from the OT Shim side (that's also fine by me).

Author: Thanks. Updated the description. Please review.
